lib/loader/serialize/serial_segment.cpp

```diff
@@ -4,32 +4,192 @@ namespace WasmEdge {
+  if (Seg.getIdx() != 0) {
+    serializeU32(Seg.getIdx(), Result);
+    if (Seg.getMode() == AST::ElementSegment::ElemMode::Active) {
+      Mode |= 0x02;
+    } else {
+      return logSerializeError(ErrCode::Value::Unreachable,
+                               ASTNodeAttr::Seg_Element);
+    }
+  }
+
+  // Serialize OffExpr.
+  if (Seg.getMode() == AST::ElementSegment::ElemMode::Active) {
+    if (auto Res = serializeExpression(Seg.getExpr(), Result); unlikely(!Res)) {
+      spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Expression));
+      return Unexpect(Res);
+    }
+  }
+
+  if (Seg.getRefType() == RefType::FuncRef) {
+    if (Mode != 0x00) {
+      // Serialize ElemKind.
+      Result.push_back(0x00);
+    }
+  } else {
+    // Serialize RefType.
+    if (auto Res = Conf.checkRefTypeProposals(Seg.getRefType(),
+                                              ASTNodeAttr::Seg_Element);
+        unlikely(!Res)) {
+      return Unexpect(Res);
+    }
+    Result.push_back(static_cast<uint8_t>(Seg.getRefType()));
+  }
+
+  // Distinguish between FuncIdx and Expr.
+  if (Seg.getInitExprs().size() != 0) {
+    auto IsExpr = false;
+    for (auto Expr : Seg.getInitExprs()) {
+      if (Expr.getInstrs().size() != 2 ||
+          Expr.getInstrs().at(0).getOpCode() != OpCode::Ref__func) {
+        IsExpr = true;
+        break;
+      }
+    }
+    if (IsExpr) {
+      Mode |= 0x04;
+    }
+  }
+
+  //  Mode > 0 cases are for BulkMemoryOperations or ReferenceTypes proposal.
+  if (Mode > 0 && !Conf.hasProposal(Proposal::BulkMemoryOperations) &&
+      !Conf.hasProposal(Proposal::ReferenceTypes)) {
+    return Conf.logNeedProposal(ErrCode::Value::ExpectedZeroByte,
+                                Proposal::BulkMemoryOperations,
+                                ASTNodeAttr::Seg_Element);
+  }
+
+  serializeU32(static_cast<uint32_t>(Seg.getInitExprs().size()), Result);
+  for (auto Expr : Seg.getInitExprs()) {
+    if (Mode & 0x04) {
+      // Serialize vec(expr).
+      if (auto Res = serializeExpression(Expr, Result); unlikely(!Res)) {
+        spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Seg_Element));
+        return Unexpect(Res);
+      }
+    } else {
+      // Serialize vec(FuncIdx).
+      for (auto Instr : Expr.getInstrs()) {
+        if (Instr.getOpCode() == OpCode::End) {
+          break;
+        }
+        serializeU32(Instr.getTargetIndex(), Result);
+      }
+    }
+  }
+
+  Result.insert(Result.begin(), Mode);
+  OutVec.insert(OutVec.end(), Result.begin(), Result.end());
+  return {};
 }
 
 // Serialize code segment. See "include/loader/serialize.h".
-void Serializer::serializeSegment(const AST::CodeSegment &,
-                                  std::vector<uint8_t> &) {
+Expect<void> Serializer::serializeSegment(const AST::CodeSegment &Seg,
+                                          std::vector<uint8_t> &OutVec) {
   // Code segment: size:u32 + locals:vec(u32 + valtype) + body:expr.
-  // TODO
+  serializeU32(Seg.getSegSize(), OutVec);
```

Comment:
Cannot serialize the code segment size here. You should refer to the serialized size bellow because you cannot ensure the serialized size is the same as the recorded size in the code segment due to the leb128 encoding.

Reply:
Done